### PR TITLE
FALCON-720. check for wf status rather than api status

### PR DESCRIPTION
**Pull request by sandeepSamudrala**

### CODE REVIEW

File: prism/src/main/java/org/apache/falcon/service/EntitySLAAlertService.java [hunk 1]
```diff
@@ -112,7 +112,7 @@ void processSLACandidates(){
         if (pendingInstanceBeanList == null || pendingInstanceBeanList.isEmpty()){
             return;
         }
-        LOG.trace("In processSLACandidates :" + pendingInstanceBeanList.size());
+        LOG.info("In processSLACandidates :" + pendingInstanceBeanList.size());
```
Comment (sandeepSamudrala) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAAlertService.java]: Was testing. I will revert this change

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 2]
```diff
@@ -353,16 +326,19 @@ public void destroy() throws FalconException {
         public void run() {
             try {
                 if (MONITORING_JDBC_STATE_STORE.getAllMonitoredEntities().size() > 0) {
-                    checkPendingInstanceAvailability(EntityType.FEED.toString());
-                    checkPendingInstanceAvailability(EntityType.PROCESS.toString());
+                    checkPendingInstanceAvailability();
 
                     // add Instances from last checked time to 10 minutes from now(some buffer for status check)
                     Date newCheckPointTime = new Date(now().getTime() + lookAheadWindowMillis);
                     addPendingEntityInstances(EntityType.FEED.toString(), null, newCheckPointTime);
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Can we change this also on the lines of checkPendingInstanceAvailability(), so we don't call this method twice?

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 3]
```diff
@@ -436,19 +417,20 @@ private boolean checkEntityInstanceAvailability(String entityName, String cluste
...
         try {
-            if (entityType.equals(EntityType.PROCESS.toString())){
+            if (entityType.equalsIgnoreCase(EntityType.PROCESS.toString())){
                 LOG.trace("Checking instance availability status for entity:{}, cluster:{}, "
                         + "instanceTime:{}", entity.getName(), clusterName, nominalTime, entityType);
                 AbstractWorkflowEngine wfEngine = WorkflowEngineFactory.getWorkflowEngine();
                 InstancesResult instancesResult = wfEngine.getStatus(entity, nominalTime, nominalTime, null, null);
-                if (instancesResult.getStatus().equals(APIResult.Status.SUCCEEDED)){
+                if (instancesResult.getInstances().length > 0
+                        && instancesResult.getInstances()[0].status.equals(InstancesResult.WorkflowStatus.SUCCEEDED)){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Should we also consider manual kills (IGNORE/KILLED) also as availability?

File: prism/src/test/java/org/apache/falcon/service/EntitySLAAlertServiceTest.java [hunk 4]
```diff
@@ -172,6 +172,7 @@ public static void processSLACandidateProcess() throws FalconException, Interrup
 
 
         Thread.sleep(10*1000);
+        // process is not scheduled.
         Assert.assertTrue(monitoringJdbcStateStore.getEntityAlertInstance("test-process", "test-cluster", dateOne,
```
Comment (pallavi-rao) [inline prism/src/test/java/org/apache/falcon/service/EntitySLAAlertServiceTest.java]: If process is not scheduled, why would isSLAHighMissed be true?

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java (hunk 2 above)
Comment (sandeepSamudrala) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Ack. Done.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java (hunk 3 above)
Comment (sandeepSamudrala) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: User could have ignored the instance for some faulty code or input, which the user might rerun later. Not taking it in the pending instances could mean user might lost the track of the instance.

File: prism/src/test/java/org/apache/falcon/service/EntitySLAAlertServiceTest.java (hunk 4 above)
Comment (sandeepSamudrala) [inline prism/src/test/java/org/apache/falcon/service/EntitySLAAlertServiceTest.java]: I will revert this change . Its supposed to throw exception as per the code. I will debug through this.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAAlertService.java (hunk 1 above)
Comment (sandeepSamudrala) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAAlertService.java]: Reverting it. Was testing it.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 5]
```diff
@@ -380,29 +351,32 @@ private void addPendingInstances(String entityType, Entity entity,
...
-                if (currentClusters.contains(entityCluster.getName())) {
-                    List<Date> instances = EntityUtil.getEntityInstanceTimesInBetween(entity, entityCluster.getName(),
-                            lastMonitoredInstanceTime, newCheckPointTime);
-                    addPendingInstances(entityType, entity, entityCluster.getName(), instances);
-                    // update last monitored time with the new checkpoint time
-                    MONITORING_JDBC_STATE_STORE.updateLastMonitoredTime(entityName, entityType,
-                            new Date(newCheckPointTime.getTime() + MINUTE_DELAY));
+            String entityType = monitoredEntityBean.getEntityType();
+            if (EntityType.FEED.name().equals(entityType)
+                    || !workflowEngine.isSuspended(EntityUtil.getEntity(entityType, entityName))) {
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: You should be using the equivalent of AbstractEntityManager.getStatus.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 6]
```diff
@@ -138,26 +116,40 @@ public void onAdd(Entity entity) throws FalconException {
...
-                        MONITORING_JDBC_STATE_STORE.putMonitoredEntity(process.getName(),
-                                EntityType.PROCESS.toString(), new Date(now().getTime() + MINUTE_DELAY));
-                        List<Date> instances = EntityUtil.getEntityInstanceTimesInBetween(entity, cluster,
-                                getStartTime(entity, cluster), now());
-                        addPendingInstances(entity.getEntityType().name().toLowerCase(), entity, cluster, instances);
+                        if (isEntityUpdated) {
+                            MONITORING_JDBC_STATE_STORE.putMonitoredEntity(process.getName(),
+                                    EntityType.PROCESS.toString(), new Date(now().getTime() + MINUTE_DELAY));
+                        } else {
+                            if (workflowEngine.isActive(entity)) {
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Prefer if AbstractEntityManager.getStatus kind of method here too.

File: common/src/main/java/org/apache/falcon/update/UpdateHelper.java [hunk 7]
```diff
@@ -46,13 +46,13 @@
 
     private static final String[] FEED_FIELDS = new String[]{"partitions", "groups", "lateArrival.cutOff",
                                                              "schema.location", "schema.provider", "tags",
-                                                             "group", "owner", "permission", };
```
Comment (sandeepSamudrala) [inline common/src/main/java/org/apache/falcon/update/UpdateHelper.java]: No. it checks at main class level. It filters out whole sla tag and ignores the changes in it.

**Pull request merged**